Support named parameters in MySQLi connection->prepare()

MalteWunsch opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues |

#### Summary

According to the [documentation](https://www.doctrine-project.org/projects/doctrine-dbal/en/2.10/reference/data-retrieval-and-manipulation.html#dynamic-parameters-and-prepared-statements), you can use named parameters in `$connection->prepare($query)`. This PR tests and adds support for that with MySQLi connections.

MalteWunsch committed:
Test MySQLi connection->prepare() with named parameters

MalteWunsch committed:
Support named parameters in MySQLi connection->prepare()

MalteWunsch committed:
Fix DocBlocks for static analysis and code style

SenseException commented:
Is this currently still WIP? It seems like the new data provider is breaking tests.

MalteWunsch commented:
Right, but I'm not sure how to solve that problem.

My intention was to add named parameters support in MySQLi connection->prepare() only, and with MySQLi the tests run well. But the new tests are run with many other drivers / database connections which do not support named parameters there, resulting in them failing.

I don't think I'm able to add support for named parameters to the other drivers, as I don't have these databases at hand.

Should I limit the test to MySQLi and if so: how do I do that? Maybe the test class annotation `@group DDC-1372` indicates in one way or the other which databases are used?

SenseException reviewed (changes requested):
> Should I limit the test to MySQLi

You'll find that there are such limitations in the tests if the tested code is only supporting one specific driver. You can find one on the `setUp()` method of `Doctrine\Tests\DBAL\Functional\Driver\Mysqli\ConnectionTest` where it skips every database that isn't using the mysqli driver or doesn't have the mysqli extension. Please also put your test into the same directory and rename it to `MysqliStatementTest`.

MalteWunsch committed:
Revert "Fix DocBlocks for static analysis and code style"

This reverts commit 6688ff3796137dd1be0796083e1a336995b4d5ca.

MalteWunsch committed:
Revert "Test MySQLi connection->prepare() with named parameters"

This reverts commit b89d51e5fdf98d9fed8e633b57ed63a26108d484.

MalteWunsch committed:
Functional test for named parameters in MySQLiStatement

MalteWunsch committed:
Obey code style

MalteWunsch committed:
Fix docBlock for static analysis

MalteWunsch committed:
Simplify execute() method

Scrutinizer found it to be too long and to have too many conditions and hence to be too complex. Reduce length and complexity by extracting methods and using guards.

MalteWunsch commented:
Thanks for the hints!

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -54,6 +60,9 @@ class MysqliStatement implements IteratorAggregate, Statement
     protected $types;
 
+    /** @var array<string, array<int, int>> maps parameter names to their placeholder number(s). */
+    protected $placeholderNamesToNumbers = [];
```
Let's go private by default, regardless of other properties
```suggestion
    private $placeholderNamesToNumbers = [];
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -97,6 +107,35 @@ public function __construct(mysqli $conn, $prepareString)
+     *
+     * @return string
+     */
+    private function convertNamedToPositionalPlaceholders($query)
```
Type declarations of docblocks please
```suggestion
     */
    private function convertNamedToPositionalPlaceholders(string $query) : string
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -97,6 +107,35 @@ public function __construct(mysqli $conn, $prepareString)
+
+            $placeholderPositionInShortenedQuery = $placeholderPosition - $numberOfCharsQueryIsShortenedBy;
+            $placeholderNameLength               = strlen($placeholderName);
+            $query                               = substr($query, 0, $placeholderPositionInShortenedQuery) . '?' . substr($query, ($placeholderPositionInShortenedQuery + $placeholderNameLength + 1));
```
Let's use linebreaks and remove unneeded parenthesis
```suggestion
            $query                               = substr($query, 0, $placeholderPositionInShortenedQuery) . '?' . 
                substr($query, $placeholderPositionInShortenedQuery + $placeholderNameLength + 1);
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
+    /**
+     * Converts an array of named parameters, e.g. ['id' => 1, 'foo' => 'bar'] to the corresponding array with
+     * positional parameters referring to the prepared query, e.g. [1 => 1, 2 => 'bar', 3 => 'bar'] for a prepared query
+     * like "SELECT id FROM table WHERE foo = :foo and baz = :foo".
```
```suggestion
     * such as "SELECT id FROM table WHERE foo = :foo and baz = :foo".
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
+     *
+     * @param array<int|string, mixed>|null $params
+     *
+     * @return mixed[]|null more specific: array<int, mixed>, I just don't know an elegant way to convince phpstan
```
Can you elaborate? What's wrong with phpstan here? Please reproduce the issue on https://phpstan.org/ so that we can give it a try.

MalteWunsch replied:
Minimal example for the phpstan playground:

```
<?php declare(strict_types = 1);

class Example {
    /**
     * @param array<int|string, mixed> $params
     *
     * @return array<int, mixed>
     */
    private function convertNamedToPositionalParamsIfNeeded(array $params): array
    {
        if ($this->arrayHasOnlyIntegerKeys($params)) {
            return $params;
        }

        // removed logic
        return [];
    }
	
    /**
     * @param array<int|string, mixed> $array
     *
     * @return bool
     */
    private function arrayHasOnlyIntegerKeys(array $array): bool
    {
        return count(array_filter(array_keys($array), 'is_int')) === count($array);
    }	
}
```

Then the error is the following: "Method Example::convertNamedToPositionalParamsIfNeeded() should return array<int, mixed> but returns array<int|string, mixed>."

I don't know an elegant way to convince phpstan that the keys in the array are actually  integers. I tried various things like

```
foreach (array_keys($array) as $key) {
    assert(is_int($key));
}
```

without success. The only way I found was creating a new array explicitly with integers as keys, but obviously that's not a good idea with respect to performance/memory.

One could use `array<int|string, mixed>` as return type docblock, being more specific than mixed[]. But for me, this level of exactness felt a little bit like a sham.

Do you have an oppinion - or even better: a solution on that?

greg0ire replied:
Here is a solution you can use https://phpstan.org/r/c66bb446-e691-4902-a013-e21c9dd9f08f

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
+     *
+     * @return bool
+     */
+    private function arrayHasOnlyIntegerKeys(array $array)
```
```suggestion
     */
    private function arrayHasOnlyIntegerKeys(array $array) : bool
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
+     *
+     * @return mixed[]|null more specific: array<int, mixed>, I just don't know an elegant way to convince phpstan
+     */
+    private function convertNamedToPositionalParamsIfNeeded(?array $params = null)
```
```suggestion
    private function convertNamedToPositionalParamsIfNeeded(?array $params = null) : array
```

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
+     */
+    private function convertNamedToPositionalParamsIfNeeded(?array $params = null)
+    {
+        if ($params === null || count($params) === 0) {
```
The case of `[]` does not need to be singled out IMO.
```suggestion
        if ($params === null) {
```

MalteWunsch replied:
The idea was to deal with any kind of emptyness first, but the empty array behaves well in the following code, too. So I'm happy to go either way.

greg0ire commented on lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php:
```diff
@@ -450,4 +519,28 @@ public function getIterator()
         return new StatementIterator($this);
     }
+
+    private function initializeColumnNamesIfNeeded()
```
```suggestion
    private function initializeColumnNamesIfNeeded() : void
```

greg0ire commented on tests/Doctrine/Tests/DBAL/Functional/Driver/Mysqli/MysqliStatementTest.php:
```diff
@@ -0,0 +1,142 @@
+                ['foo' => '"\''],
+                [],
+            ],
+        ];
```
What happens if there are more parameters than needed in the query? Or when there are fewer than needed (an exception I hope).

MalteWunsch replied:
Too few parameters:

```
Doctrine\DBAL\DBALException: An exception occurred while executing 'SELECT * FROM mysqli_statement_test f WHERE f.foo = :foo':

Wrong parameter count for mysqli_stmt::bind_param()
```

Too many parameters:

```
Doctrine\DBAL\DBALException: An exception occurred while executing 'SELECT * FROM mysqli_statement_test f WHERE f.foo = :foo' with params [1, 2]:

Undefined index: bar
```

where `bar` is the name of the superfluous named parameter.

Are these exceptions okay? Do you want a test for them?

greg0ire replied:
I will answer once and if https://github.com/doctrine/dbal/pull/3891#issuecomment-606299763 is resolved

MalteWunsch committed:
Use private scope by default

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Use return type declaration

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Improve readability

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Avoid SQL keyword in running text

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Declare return type

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Declare return type

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Remove guard for empty arrays, as they can pass

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Declare return type

Co-Authored-By: Grégoire Paris <postmaster@greg0ire.fr>

MalteWunsch committed:
Fix NULLability of return type + phpcs

morozov commented:
@MalteWunsch named parameters and their conversion to the positional ones are supported at the wrapper connection level, not at the driver connection. Why do you need to implement them at the driver level?
